### PR TITLE
feat(geoip transform): ASN/ISP database support

### LINKED ISSUE #1372: Support GeoIP ASN in `geoip` transform
Support for ASN ?
https://dev.maxmind.com/geoip/geoip2/geolite2-asn-csv-database/
Get isp and Network of ip.

### PR DESCRIPTION
**Pull request by markonen**
This adds support for the GeoIP2-ISP and GeoLite2-ASN databases in the `geoip` transform. The database type is autodetected by the `database_type` metadata field in the binary database file; if an ISP/ASN database is not detected, the City database (and all of the existing logic) will run as usual. No changes to configuration options were added.

I'm slightly unhappy about the way the documentation page now includes an example log event section that includes fields from both the City and ISP databases, even though you only ever get one or the other (unless you run the transform twice). But changing the machinery that generates this example was a bridge too far for me 🙃

Resolves #1372 

Signed-off-by: Marko Karppinen <marko@richie.fi>

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice work. We should be able to classify the output fields. I'll see if I can clean that up.

**Comment by markonen** [comment]
Will fix the lint issue later today

**Comment by markonen** [comment]
Fixed a couple of minor issues in the docs. Not sure where/how the URLs should be included after the doc site split, though.

**Comment by binarylogic** [comment]
Thanks! I added a `.meta/links.toml` file in this PR to add link references, so you should be able to add them there.

**Comment by binarylogic** [comment]
Hi @markonen this is looking good.

Since you're somewhat familiar with the `geoip` code, would you be interested in helping out with https://github.com/timberio/vector/issues/3395? We're working on observability for Vector itself (`top` subcommand and admin web UI) and adding these events will help to observe this transform. If so, a separate PR would do. We'll get this reviewed shortly.

File: src/transforms/geoip.rs [hunk 1]
```diff
@@ -146,19 +186,48 @@ impl Transform for Geoip {
...
+                "{}.autonomous_system_number",
+                target_field
+            )))
+            .is_none()
+            {
+                e.insert(
+                    Atom::from(format!("{}.autonomous_system_number", target_field)),
+                    Value::from(0),
+                );
+            }
```
Comment (fanatid) [inline src/transforms/geoip.rs:203]:
```suggestion
            e.try_insert(&Atom::from(format!("{}.autonomous_system_number", target_field)), Value::from(0));
```

Same for other `get(..).is_none()` + `insert`.

File: src/transforms/geoip.rs [hunk 2]
```diff
@@ -146,19 +186,48 @@ impl Transform for Geoip {
...
+                format!("{}.country_code", target_field),
+                format!("{}.continent_code", target_field),
+                format!("{}.timezone", target_field),
+                format!("{}.latitude", target_field),
+                format!("{}.longitude", target_field),
+                format!("{}.postal_code", target_field),
+            ];
+            for field in geoip_fields.iter() {
+                let e = event.as_mut_log();
+                if e.get(&Atom::from(field.to_string())).is_none() {
```
Comment (fanatid) [inline src/transforms/geoip.rs:228]: It's should be possible to use `field` without `to_string()`, `into_iter()` instead `iter()` should help.

**Comment by markonen** [comment]
@fanatid thank you, I reworked the part that fills in the empty values in a way that's hopefully clearer. Could you take a look?

One thing I noticed is that `try_insert()` seems to be noisy in the `make test` output:

```
Looking for Atom('geo.longitude' type=dynamic)
Looking for Atom('geo.country_code' type=dynamic)
Looking for Atom('geo.postal_code' type=dynamic)
Looking for Atom('geo.latitude' type=dynamic)
Looking for Atom('geo.city_name' type=dynamic)
Looking for Atom('geo.continent_code' type=dynamic)
Looking for Atom('geo.timezone' type=dynamic)
test transforms::geoip::tests::geoip_city_lookup_no_results ... ok
test transforms::geoip::tests::geoip_city_lookup_success ... ok
Looking for Atom('geo.autonomous_system_number' type=dynamic)
Looking for Atom('geo.autonomous_system_organization' type=dynamic)
Looking for Atom('geo.isp' type=inline)
Looking for Atom('geo.organization' type=dynamic)
test transforms::geoip::tests::geoip_isp_lookup_no_results ... ok
```

Is this bad?

**Comment by fanatid** [comment]
But this output going from test `geoip_isp_lookup_no_results` which you added? :slightly_smiling_face:

**Comment by markonen** [comment]
@fanatid oops, sorry, I was totally convinced that this output wasn't there before so I was sure it was emitted by your recommended `try_insert()` rather than the test itself 🙈

**Comment by markonen** [comment]
I took another stab at this based on comments from @bruceg. I think it's cleaner now, but have a look.

@binarylogic sorry, can't see that `links.toml` file anywhere?

**Comment by markonen** [comment]
@binarylogic any ideas about this docs failure?

```
Failed! * Message: Additional properties are not allowed ('how_it_works' was unexpected)
```

**Comment by binarylogic** [comment]
@markonen all fixed. Once tests pass we'll merge. Thanks again for your help.

**Comment by binarylogic** [comment]
The shutdown error is not related to this PR, so I'll be merging.

**Pull request merged**